lib/propolis/src/enlightenment/hyperv/overlay.rs

```diff
@@ -427,6 +476,126 @@ impl ManagerInner {
+
+        migrate::HyperVOverlaysV1 { overlays: map }
+    }
+
+    fn import(
+        &mut self,
+        manager: &Arc<OverlayManager>,
+        payload: migrate::HyperVOverlaysV1,
+    ) -> Result<Vec<OverlayPage>, OverlayError> {
+        // Take each serialized (PFN, overlay set) tuple and convert it back
+        // into an `OverlaySet` that can be inserted into the map. Because the
+        // input payload came over the wire, it is not strictly guaranteed that
+        // it came from a Propolis that upheld all of the overlay set
+        // invariants, so these need to be rechecked and an error returned if
+        // one is violated. These include:
+        //
+        // - All overlay page contents must in fact be 4K in size.
+        // - All PFNs must be valid 52-bit page numbers for 4K pages.
+        // - Each overlay kind must appear at most once in each overlay set.
+        self.overlays = payload
+            .overlays
+            .into_iter()
+            .map(|(pfn, set)| -> Result<(Pfn, OverlaySet), OverlayError> {
+                let original_contents = OverlayContents::try_from(
+                    set.original_contents,
+                )
+                .map_err(|len| {
+                    OverlayError::ImportedContentsNotPageSized { pfn, len }
+                })?;
+
+                let pfn = Pfn::new(pfn)
+                    .ok_or(OverlayError::ImportedPfnTooLarge(pfn))?;
+
+                let pending = set
+                    .pending
+                    .into_iter()
+                    .map(|(kind, contents)| -> Result<_, OverlayError> {
+                        if kind == set.active {
+                            return Err(OverlayError::KindInUse(
+                                pfn.into(),
+                                kind.into(),
+                            ));
+                        }
+
+                        let contents = OverlayContents::try_from(contents)
+                            .map_err(|len| {
+                                OverlayError::ImportedContentsNotPageSized {
+                                    pfn: pfn.into(),
+                                    len,
+                                }
+                            })?;
+
+                        Ok((OverlayKind::from(kind), contents))
+                    })
+                    .collect::<Result<BTreeMap<_, _>, OverlayError>>()?;
+
+                Ok((
+                    pfn,
+                    OverlaySet {
+                        original_contents,
+                        active: set.active.into(),
+                        pending,
+                    },
+                ))
+            })
+            .collect::<Result<BTreeMap<Pfn, OverlaySet>, OverlayError>>()?;
+
+        // Now that the tracking tables have been reconstructed, create a set of
+        // overlay page descriptors to return to the caller. The caller will
+        // audit these to make sure they align with other enlightenment state.
```

Comment:
Wait, sorry, what exactly are you referring to here with "the caller will audit these..."? Is that just the check that the hypercall page's PFN matches the hypercall MSR's PFN?

Reply:
More or less (there's also a check for whether the Hyper-V stack and the overlay manager even agree that an overlay should exist). I clarified this comment a bit in 4f99c85.